### PR TITLE
[RRIO] [Cache] Stub CacheRead transform

### PR DESCRIPTION
**Pull request by damondouglas**
This PR addresses #28907 via stubbing the `class CacheRead<RequestT, ResponseT> extends PTransform<PCollection<RequestT>, Result<RequestT, ResponseT>>`

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by damondouglas** [comment]
assign set of reviewers

**Review by ahmedabu98 — CHANGES_REQUESTED** [review]

File: sdks/java/io/rrio/src/main/java/org/apache/beam/io/requestresponseio/CacheRead.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+  public Result<RequestT, ResponseT> expand(PCollection<RequestT> input) {
+    return Result.of(
+        new TupleTag<KV<RequestT, ResponseT>>() {}, PCollectionTuple.empty(input.getPipeline()));
+  }
+
+  /**
+   * The {@link Result} of reading RequestT {@link PCollection} elements yielding ResponseT {@link
+   * PCollection} elements.
+   */
+  static class Result<RequestT, ResponseT> implements POutput {
```
[the remark this answers is not in the document]
Comment (damondouglas) [reply sdks/java/io/rrio/src/main/java/org/apache/beam/io/requestresponseio/CacheRead.java:78]: As we talked about, the only alternative here is a PCollectionTuple which is rather cumbersome to use and check for existence of a tag/PCollection. I'll go ahead and resolve this. Please re-open if you would like to re-visit.

File: sdks/java/io/rrio/src/main/java/org/apache/beam/io/requestresponseio/CacheRead.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    private final Pipeline pipeline;
+    private final TupleTag<KV<RequestT, ResponseT>> responseTag;
+    private final PCollection<KV<RequestT, ResponseT>> responses;
+    private final PCollection<ApiIOError> failures;
+
+    private Result(TupleTag<KV<RequestT, ResponseT>> responseTag, PCollectionTuple pct) {
+      this.pipeline = pct.getPipeline();
+      this.responseTag = responseTag;
+      this.responses = pct.get(responseTag);
+      this.failures = pct.get(FAILURE_TAG);
```
[the remark this answers is not in the document]
Comment (damondouglas) [reply sdks/java/io/rrio/src/main/java/org/apache/beam/io/requestresponseio/CacheRead.java:94]: We do not know the type of the PCollection tagged by the responseTag but we do the FAILURE_TAG. Java doesn't allow static Generic types. This is why the FAILURE TAG is static final and the responseTag is only final. I will go ahead and resolve this but please re-open if you would like to re-visit.

File: sdks/java/io/rrio/src/main/java/org/apache/beam/io/requestresponseio/CacheRead.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (johnjcasey) [reply sdks/java/io/rrio/src/main/java/org/apache/beam/io/requestresponseio/CacheRead.java:78]: I gave bad advice to Ahmed here, and I misread POutput as PDone. POutput is fine, PDone is discouraged. sorry about the confusion

[the remark this answers is not in the document]
Comment (ahmedabu98) [reply sdks/java/io/rrio/src/main/java/org/apache/beam/io/requestresponseio/CacheRead.java:78]: After talking to other folks, I take back my statement. PDone is the problematic class, not POutput so this should be good to go

**Review by ahmedabu98 — APPROVED** [review]

**Pull request merged**